### ISSUE TITLE
inline or not inline ...

### ISSUE THREAD

**Issue by tpaviot** [andreasplesch/OCCToX3D#6]
... that is the question I ask to myself for a while.

Converting a STEP file to X3D can be achieved in two ways:

* generate one single X3D file with many shapes

* generate one X3D file per shape, and make them consistent at using inline. At pythonocc, that is the way I chose, because loading in the viewer can be done in an async manner, but I'm not sure it's the best choice.

**Comment by andreasplesch** [comment]
Yes, async loading is nice but a single file is easier to transport and share. For example, nbviewer for Jupyter notebooks can embed html content but cannot load from a file.

So we want both, by making functions more modular. Then a python user can pick and choose.

**Comment by tpaviot** [comment]
enabling both is indeed better

**Comment by andreasplesch** [comment]
Here is an example of one may do with the provided x3d:
https://github.com/andreasplesch/OCCToX3D/blob/conda2/notebooks/colorshapes_to_x3d.ipynb
https://mybinder.org/v2/gh/andreasplesch/OCCtoX3D/conda2?filepath=notebooks/colorshapes_to_x3d.ipynb
I could use ElementTree to combine the many LineSets in the edg inlines and in each shp inline to optimize the x3d. This will be not necessary anymore but for now it seemed useful and a good exercise of how possible it is to modify the generated x3d.

**Comment by andreasplesch** [comment]
I think I will also try to combine all the inlines with ElementTree into a single x3d.

One small problem is that DEF names need to be unique but they are not currently used for for anything.

**Comment by tpaviot** [comment]
Regarding the ```DEF```, I currently use unique ids given by the ```uuid``` python module, but it doesn't hold any semantics. But there are many other way to generate uids

**Comment by andreasplesch** [comment]
The labels of a document seem like a good candidate for uids and also worth preserving (although not a step feature).

**Comment by tpaviot** [comment]
I updated the x3dom code so that it is able to export many shapes to a single file. Here is what the output looks like.
[random_boxes_single_file.zip](https://github.com/andreasplesch/OCCToX3D/files/4626571/random_boxes_single_file.zip)

I have a question regarding the ```id``` and ```DEF``` attribute. What is the difference between them ? I noticed there is a  ```mapDEFtoId``` property that can be set either to ```true``` or ```false```, so I guess there is a close relationship between them. So far, I decide to set ```id``` and ```DEF``` to the same value.

**Comment by andreasplesch** [comment]
works fine in freeWrl:
![image](https://user-images.githubusercontent.com/6171115/81978231-3661dd00-95f9-11ea-96e4-0f7e54c5c971.png)

**Comment by andreasplesch** [comment]
The `id` and `DEF` attributes are quite similar. X3D itself does not use `id` at all, and x3dom does not need the `id` attribute. It can be provided as a convenience for web dom programming. `DEF` is necessary if `USE` instancing is desired. `DEF` works across encodings, eg. for vrml and x3d xml.

x3dom has the extra `mapDEFtoId` for Inlines because regular X3D files do not have `id` attributes. If `true` the inlined elements are assigned a new `id` attribute value which is the same as the `DEF` value if it exists. This is only useful if also `namespacename` is specified. `namespacename` causes x3dom to attach the inlined content in the dom as a child to the Inline node. In addition, `id` values of the attachad nodes are prefixed (namespaced) with the `namespacename` value. This way inlined content can then be manipulated after loading, by js.

https://doc.x3dom.org/tutorials/models/inline/index.html

https://examples.x3dom.org/example/x3dom_inlineReflection.xhtml
